File: import/cppmakelib/system/linux.py
from cppmakelib.error.config       import ConfigError
from cppmakelib.utility.decorator  import member
from cppmakelib.utility.filesystem import path, resolvable_path
import sys

class Linux:
    def __init__(self: Linux) -> None: ...
    executable_suffix: str             = ''
    object_suffix    : str             = 'o'
    static_suffix    : str             = 'a'
    dynamic_suffix   : str             = 'so'
    compiler         : resolvable_path = 'g++'
    install_dir      : path            = '/usr'

    def _check(self: Linux) -> None: ...



@member(Linux)
def __init__(self: Linux) -> None:
    self._check()

@member(Linux)
def _check(self: Linux) -> None:
    if sys.platform != 'linux':
        raise ConfigError(f'linux check failed (with sys.platform = {sys.platform})')

File: import/cppmake/system/linux.cpp
export module cppmake:system.linux;
import               :error.config;
import               :utility.filesystem;

namespace cppmake
{
    export class linux
    {
        public:
            static constexpr std::string_view name              = "linux";
            static constexpr std::string_view executable_suffix = "";
            static constexpr std::string_view object_suffix     = ".o";
            static constexpr std::string_view static_suffix     = ".a";
            static constexpr std::string_view dynamic_suffix    = ".so";
            static inline    resolvable_path  compiler_path     = "g++";
            static inline    resolvable_path  linker_path       = "ld";
            static inline    absolute_path    install_path      = "/usr";

        public:
            linux ( );
    };



    linux::linux ( )
    {
        #ifndef __linux__
            throw config_error("__linux__ is not defined");
        #endif
    }
}